internal/repos/store.go

```diff
@@ -272,6 +273,89 @@ WHERE id = %s AND NOT EXISTS (
 )
 `
 
+const listExternalServiceUserIDsByRepoIDQuery = `
+SELECT user_id FROM external_service_repos
+WHERE repo_id = %s AND user_id IS NOT NULL
+`
+
+// ListExternalServiceUserIDsByRepoID returns the users associated with a given
+// repository. These users have proven that they have read access to the
+// repository given their presence in our external_service_repos table.
+func (s *Store) ListExternalServiceUserIDsByRepoID(ctx context.Context, repoID api.RepoID) (userIDs []int32, err error) {
+	if database.Mocks.Repos.ListExternalServiceUserIDsByRepoID != nil {
+		return database.Mocks.Repos.ListExternalServiceUserIDsByRepoID(ctx, repoID)
+	}
+
+	tr, ctx := s.trace(ctx, "Store.ListExternalServiceUserIDsByRepoID")
+	tr.LogFields(
+		otlog.Int32("repo_id", int32(repoID)),
+	)
+
+	defer func(began time.Time) {
+		secs := time.Since(began).Seconds()
+		s.Metrics.ListExternalServiceUserIDsByRepoID.Observe(secs, 1, &err)
+		logging.Log(s.Log, "store.list-external-service-user-ids-by-repo-id", &err,
+			"repo-id", repoID,
+		)
+		tr.SetError(err)
+		tr.Finish()
+	}(time.Now())
+
+	q := sqlf.Sprintf(listExternalServiceUserIDsByRepoIDQuery, repoID)
+	rows, err := s.Query(ctx, q)
+	if err != nil {
+		return nil, err
+	}
+	return basestore.ScanInt32s(rows, err)
+}
+
+const listExternalServiceRepoIDsByUserIDQuery = `
+SELECT repo_id FROM external_service_repos
+WHERE user_id = %s
+`
```

Comment:
As mentioned in the TODO above, we only care about private repos here

```suggestion
const listExternalServiceRepoIDsByUserIDQuery = `
SELECT repo_id FROM external_service_repos esr
JOIN repo r ON r.id = esr.repo_id
WHERE user_id = %s
AND r.private
`
```